odoo/addons/base/models/ir_async.py

```diff
@@ -0,0 +1,242 @@
+        threading.current_thread().dbname = dbname
+
+        # make sure we can process this db
+        with db.cursor() as cr:
+            cr.execute("""
+                SELECT latest_version
+                FROM ir_module_module
+                WHERE name=%s""", ('base',))
+            (version,) = cr.fetchone()
+            if version is None:
+                _logger.warning(
+                    "Skipping database %s because of modules to "
+                    "install/upgrade/remove.", dbname)
+                return
+            if version != BASE_VERSION:
+                _logger.warning(
+                    "Skipping database %s as its base version is not %s.",
+                    dbname, BASE_VERSION)
+                return
+
+        # process all jobs enqueued in this db
+        while True:
+            with db.cursor() as manager_cr:
+
+                # acquire job
+                manager_cr.execute("""
+                    SELECT *
+                    FROM ir_async
+                    WHERE state = 'created'
+                    LIMIT 1
+                    FOR UPDATE SKIP LOCKED
+                    """)
+                job = manager_cr.dictfetchone()
+                if job is None:
+                    break
+
+                # process job
+                with api.Environment.manage():
+                    with db.cursor() as job_cr:
+                        cls._process_job(job_cr, job)
+
+                # update job
+                manager_cr.execute("""
+                    UPDATE ir_async
+                    SET state=%s,
+                        payload=%s,
+                        write_date=%s
+                    WHERE ID = %s""", (
+                        job['state'],
+                        json.dumps(job['payload']),
+                        datetime.datetime.utcnow(),
+                        job['id'],
+                    ))
+
+    @staticmethod
+    def _process_job(job_cr, job):
+        """
+        Process one job, restore environment and recordset then safely
+        call the desired model method.
+        """
+        for json_field in ('res_ids', 'args', 'kwargs', 'context'):
+            job[json_field] = json.loads(job[json_field])
+        job['context']['async_traceback'] = job['traceback']
+        job['context']['async_job_id'] = job['id']
+        env = api.Environment(job_cr, job['user_id'], job['context'], job['super_user'])
+        ir_async = env['ir.async']
+
+        ir_async._pre_process(job)
+
+        try:
+            _logger.info('Calling "%s" on "%s"', job['method'], job['res_model'])
+            records = env[job['res_model']].browse(job['res_ids'])
+            result = getattr(records, job['method'])(*job['args'], **job['kwargs'])
+            json.dumps(result)  # ensure result is serializable
+            records.flush()
+        except Exception as exc:
+            job['state'] = 'failed'
+            job['payload'] = ir_async._handle_exception(exc)
```

Comment:
Shouldn't we use a savepoint in case the transaction is in failed state?
This is a problem, because the caller will not be able to commit.
We need a test case for this.

Reply:
So...

```
savepoint()
try:
    ...
except Exception:
    rollback()
else:
    pass  # just let the context manager commit```

And an associated test where I have some sort of async job that do `self._cr.commit()` then raise an exception ? The valid case should the commited values are rollbacked thanks to the savepoint ?